### PR TITLE
fix: update currency validation

### PR DESCRIPTION
**Pull request by leoslr**
## Description of the changes

Improve the way we validate currencies on request creation.

First step for https://app.asana.com/0/1203350802794948/1203507113157162

### CODE REVIEW

File: packages/currency/test/currencyManager.test.ts [hunk 1]
```diff
@@ -591,6 +591,95 @@ describe('CurrencyManager', () => {
     });
   });
 
+  describe.only('Validate currencies', () => {
```
Comment (benjlevesque) [inline packages/currency/test/currencyManager.test.ts:594]:
```suggestion
  describe('Validate currencies', () => {
```

File: packages/currency/test/currencyManager.test.ts [hunk 2]
```diff
@@ -591,6 +591,95 @@ describe('CurrencyManager', () => {
...
+          },
+          label: 'ERC777 currency',
+        },
+      ])('Should validate $label', ({ currency }) => {
+        const result = CurrencyManager.validateCurrency(currency);
+        expect(result).toBe(true);
+      });
+    });
+
+    describe('Valid cases', () => {
```
Comment (benjlevesque) [inline packages/currency/test/currencyManager.test.ts:650]:
```suggestion
    describe('Invalid cases', () => {
```

File: packages/currency/src/currencyManager.ts [hunk 3]
```diff
@@ -228,28 +228,53 @@ export class CurrencyManager<TMeta = unknown> implements ICurrencyManager<TMeta>
...
+    if (currency.type === RequestLogicTypes.CURRENCY.ISO4217) {
+      throw new Error(`Could not validate an address for an ISO4217 currency`);
+    }
+    return this.validateAddressWithNetworkAndType(address, currency.type, currency.network);
+  }
+
+  /**
+   * Validate the correctness of a Storage Currency
+   */
+  static validateCurrency(currency: StorageCurrency): boolean {
```
Comment (benjlevesque) [inline packages/currency/src/currencyManager.ts:240]:
```suggestion
  static isValidCurrency(currency: StorageCurrency): boolean {
```

File: packages/currency/src/currencyManager.ts [hunk 4]
```diff
@@ -228,28 +228,53 @@ export class CurrencyManager<TMeta = unknown> implements ICurrencyManager<TMeta>
...
+      currency.value,
+      currency.type,
+      currency.network as string,
+    );
+  }
+
+  /**
+   * Validates the correctness of an address based on a currency type and a network
+   */
+  static validateAddressWithNetworkAndType(
```
Comment (benjlevesque) [inline packages/currency/src/currencyManager.ts:257]: maybe this should be an overload of `validateAddress` (which could be renamed 
`isValidAddress`)

File: packages/request-client.js/test/index.test.ts [hunk 5]
```diff
@@ -1532,6 +1532,44 @@ describe('request-client.js', () => {
     });
   });
 
+  it.only('cannot create ERC20 address based requests with invalid currency', async () => {
```
Comment (olivier7delf) [inline packages/request-client.js/test/index.test.ts:1535]:
```suggestion
  it('cannot create ERC20 address based requests with invalid currency', async () => {
```

File: packages/currency/src/currencyManager.ts [hunk 6]
```diff
@@ -228,28 +228,53 @@ export class CurrencyManager<TMeta = unknown> implements ICurrencyManager<TMeta>
    * Throws if the currency is an ISO4217 currency.
    */
   static validateAddress(address: string, currency: CurrencyInput): boolean {
```
Comment (olivier7delf) [inline packages/currency/src/currencyManager.ts:230]: regarding the naming what do you think about:
validateAddress into validateCryptoCurrencyAddress, as it validates tokens or native tokens currencies. (no fiat)
validateAddressWithNetworkAndType into validateCurrencyAddress

File: packages/currency/src/currencyManager.ts (hunk 4 above)
Comment (leoslr) [reply packages/currency/src/currencyManager.ts]: I removed this function and made the input of `validateAddress` an union (`ICurrencyInput | StorageCurrency`) instead.

**Review by benjlevesque — APPROVED** [review]
Great!

**Comment by leoslr** [comment]
@olivier7delf @benjlevesque 
For the naming, the function `validateAddressWithNetworkAndType` has been removed.
So we only have `validateAddress` and `validateCurrency`. Is it worth renaming them `isValidAddress` and `isValidCurrency`, knowing that it will require to update several part of the stack as `validateAddress` is used at several places ?

**Review by olivier7delf — APPROVED** [review]

**Pull request merged**